### ISSUE TITLE
wasm-split utilizing DCE

### ISSUE THREAD

**Issue by ravisumit33** [emscripten-core/emscripten#20178]
We have a complex app for which currently supported `profile guided` way of wasm split is not possible. However, we were thinking of utilizing DCE to get all functions to keep.

Our app has exposed some APIs from the c++ code. We want all functions related to only some of the APIs to present in primary module, rest all can go in the secondary module. We created two builds, one with all APIs(full build) and one with APIs that are only supposed to go in primary module(profile build). Then, with the help of emitted symbol map we listed all the functions that need to be kept and passed it for `--keep-funcs` wasm-split cmd args. Split done this way worked fine with no optimizations (e.g. `-O0`) but with `-O3` we started to notice some functions required for the primary module APIs that are there in the full build but not in the profile build. We also tried with building full build with `-O3` and profile build with `-O0` to remove inlining cases but issue was still not solved.

Do you think this approach to wasm-split is feasible? If yes, what can be done to solve the above mentioned issue?

**Comment by ravisumit33** [comment]
@tlively @sbc100 Thoughts?

**Comment by sbc100** [comment]
It sounds like an optimizer is making different decisions in your two builds.   That could either be an LLVM LTO time or at binaryen/wasm-opt time I suppose.

Another way to do this might be take your list of API entry points and then do some kind of call graph analysis to find the list of all function that are transitively needed.  I'm not sure what the best tool would be to do that analysis but there would be something like that in wabt of binaryen that already exists. 

Alternatively, we could add a flag to wasm-split that tells it to transitively include all function needed by the specified list of functions.  @tlively does that sounds reasonable/easy?

**Comment by ravisumit33** [comment]
I thought of the same call graph thing and was able to create call graphs in sample programs until I came up with runtime calling of function through `call_indirect` instruction. If I have to provide the list of all the functions I want to keep to `wasm-split` tool, I want all the functions that are called either directly through `call` instruction or indireclty through `call_indirect`. Later of which, I am unable to find.

**Comment by sbc100** [comment]
How about a hybrid of the two approaches.

1. Use the "call graph" approach on the final program to get function list A
2. Use the "two builds" approach and then use the static table slot in the smaller program to generate list B
3. Combine list A list B to get the final function list to pass to wasm-split.

That way, the only way an error could occur would be if the optimizer in the "two programs" approach was somehow doing a different level of de-virtualization (which seems unlikely)?

**Comment by ravisumit33** [comment]
IIUC, the static table you are refering is `Elem` section. Does this section only contain functions are referenced indirectly? I think I have seen cases of dynamic linking with large `Elem` section.

**Comment by sbc100** [comment]
> IIUC, the static table you are refering is `Elem` section. Does this section only contain functions are referenced indirectly? 

Yes, that linker creates the elem section based on which functions are called indirectly.

> I think I have seen cases of dynamic linking with large `Elem` section.

I can imagine dynamic linking causing more function to be called indirectly, and therefore having a larger Elem section, but that above statement still holds true:  The elem section only contains functions that are called indirectly.  

Are you using dynamic linking?   Does that Elem section in the minimal/smaller program seem to contains functions that it shouldn't?

**Comment by ravisumit33** [comment]
I haven't tried smaller programs with dynamic linking but my codebase, which has total 12147 functions, has 10451 entries in the static table. And we use dynamic linking. Thus the doubt I had.

**Comment by sbc100** [comment]
I'm not sure why dynamic linking would cause an explosion in the number of indirect functions in the table, but there could well be something I'm not thinking of.

Note that C++ programs will likely have a lot of indirection function due to v-tables.  However, I don't think you should ever see the same function in the table more than once.  That should be impossible.

**Comment by ravisumit33** [comment]
<img width="555" alt="Screenshot 2023-09-07 at 1 21 06 AM" src="https://github.com/emscripten-core/emscripten/assets/19284116/652f153d-52db-40a9-bc96-3d330cf0afdb">

Above is the snapshot of `Elem` table. Here, you can see there are duplicate entries for `func[57]`

**Comment by sbc100** [comment]
Interesting.  I'm curious under what circumstances this might occur.    Can you trying linking with `EMCC_DEBUG=1` (or `EMCC_DEBUG_SAVE=1`)?   Does the duplication also occur in the wasm-ld output (`/tmp/emscripten_temp/emcc-0-base.wasm`)?

**Comment by ravisumit33** [comment]
I  built with `EMCC_DEBUG=1` and looked into `emcc-0-base.wasm` and found it has `109827` total functions in the functions section and `8243` entries in the static table. Also, I can't see the duplication here.

**Comment by tlively** [comment]
As you've discovered, different optimization choices can make this kind of splitting approach very brittle. And even if you do get it working, an Emscripten update might start making different optimization choices in LLVM or Binaryen and break you again. Since you need more control over the splitting than the profile-guided approach can give you, would you be able to manually update your code base to put the secondary API in a shared library that you can lazily dlopen when you need it?

**Comment by ravisumit33** [comment]
Both Primary and Secondary APIs share common internal code base. So, having a separate set of files to make secondary APIs, a shared library is not possible. That's why, we are trying to have wasm-split do the splitting. As mentioned above, due to complexity of the app, profile guided way is also not feasible for us. Do you find any issue with the hybrid approach suggested above?

**Comment by tlively** [comment]
The hybrid approach is probably worth a shot, but it's still possible in principle for it to fail. If the build with the secondary API removed is able to devirtualize and inline a call to a function where the full build is not able to, then the hybrid approach would incorrectly split out that function. I don't know how you could adapt this approach to guarantee that it will work correctly.

**Comment by sbc100** [comment]
Interesting.. that means that wasm-opt is introducing the duplication.  I guess via duplicate function elimination or some such.  If you look at the other .wasm files in that directory you can probably figure out which phase introduced that... although I'm not sure we really need to track that down.

**Comment by ravisumit33** [comment]
Yes, thats not really important. I tried out the hybrid approach. Looks like even for primary APIs, even static tables differ between two builds. For example, some indirect function called inside a primary API is not even there in the whole symbol map of secondary build even if secondary build is built with `-O0`. This looks a bit strange to me.

**Comment by ravisumit33** [comment]
What I can only do now is that have only one full build and create function list with the call graph of primary APIs alongwith all functions present in the static table. I am not sure if any significant size reduction we can get following this.
